library/proc_macro/src/diagnostic.rs

```diff
@@ -51,132 +132,122 @@ pub struct Diagnostic {
-        /// identified by this method's name with the given `message`.
-        #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-        pub fn $regular<T: Into<String>>(mut self, message: T) -> Diagnostic {
-            self.children.push(Diagnostic::new($level, message));
-            self
-        }
-    };
-}
-
-/// Iterator over the children diagnostics of a `Diagnostic`.
-#[derive(Debug, Clone)]
-#[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-pub struct Children<'a>(std::slice::Iter<'a, Diagnostic>);
-
-#[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-impl<'a> Iterator for Children<'a> {
-    type Item = &'a Diagnostic;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        self.0.next()
-    }
-}
-
 #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
 impl Diagnostic {
-    /// Creates a new diagnostic with the given `level` and `message`.
-    #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-    pub fn new<T: Into<String>>(level: Level, message: T) -> Diagnostic {
-        Diagnostic { level, message: message.into(), spans: vec![], children: vec![] }
+    /// Obtain a mutable reference to the last message.
+    fn last_message_mut(&mut self) -> &mut Diagnostic {
+        if self.children.is_empty() { self } else { self.children.last_mut().unwrap() }
     }
 
-    /// Creates a new diagnostic with the given `level` and `message` pointing to
-    /// the given set of `spans`.
+    /// Creates a new error diagnostic with the provided message.
     #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-    pub fn spanned<S, T>(spans: S, level: Level, message: T) -> Diagnostic
-    where
-        S: MultiSpan,
-        T: Into<String>,
-    {
-        Diagnostic { level, message: message.into(), spans: spans.into_spans(), children: vec![] }
+    #[must_use = "diagnostics do nothing unless emitted"]
+    pub fn error(message: &str) -> Self {
+        Diagnostic {
+            level: Level::Error,
+            message: message.into(),
+            spans: vec![],
+            labels: vec![],
+            children: vec![],
+        }
     }
 
-    diagnostic_child_methods!(span_error, error, Level::Error);
-    diagnostic_child_methods!(span_warning, warning, Level::Warning);
-    diagnostic_child_methods!(span_note, note, Level::Note);
-    diagnostic_child_methods!(span_help, help, Level::Help);
-
-    /// Returns the diagnostic `level` for `self`.
-    #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-    pub fn level(&self) -> Level {
-        self.level
-    }
+    // FIXME Add lint-associated warnings
 
-    /// Sets the level in `self` to `level`.
+    /// Creates a new note diagnostic with the provided message.
     #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-    pub fn set_level(&mut self, level: Level) {
-        self.level = level;
+    #[must_use = "diagnostics do nothing unless emitted"]
```

Comment:
Have we considered putting the `#[must_use]` on the `struct` definition instead?

I think that makes more sense than putting it on individual methods.

Reply:
Didn't know that was a thing. I'll change it once I'm on my laptop.